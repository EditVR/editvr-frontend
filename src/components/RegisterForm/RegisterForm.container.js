/**
 * @file RegisterForm.container.js
 * Exports a redux-connected RegisterForm component.
 */

import { connect } from 'react-redux';

import RegisterForm from './RegisterForm';

const mapDispatchToProps = dispatch => ({
  dispatch
});

const mapState = ({ user }) => ({
  user
});

export default connect(mapState, mapDispatchToProps)(RegisterForm);
