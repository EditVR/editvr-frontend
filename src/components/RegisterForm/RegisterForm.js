/**
 * @file RegisterForm.js
 * Exports a component that allows users to register for an EditVR account.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TextField, Button, withStyles } from '@material-ui/core';
import { withFormik } from 'formik';
import { string, object } from 'yup';

import { Message } from '../';
import RegisterFormStyles from './RegisterForm.style';
import {
  USER_REGISTER,
  FORM_BUTTON_LOGIN,
  FORM_BUTTON_REGISTER,
  FORM_INPUT_VALIDATION_REGEX_URL_SEGMENT
} from '../../constants';

const RegisterForm = ({
  classes,
  user: { error },
  values,
  errors,
  touched,
  isSubmitting,
  handleSubmit,
  handleChange,
  handleBlur
}) => (
  <form onSubmit={handleSubmit}>
    {error && <Message>{error}</Message>}
    <TextField
      required
      id="username"
      label="Username"
      type="text"
      className={classes.textField}
      onChange={handleChange}
      onBlur={handleBlur}
      error={!!errors.username && touched.username}
      disabled={isSubmitting}
      helperText={
        errors.username ? errors.username : 'Enter a URL-friendly username.'
      }
    />
    <TextField
      required
      id="email"
      label="Email"
      type="email"
      className={classes.textField}
      onChange={handleChange}
      onBlur={handleBlur}
      error={!!errors.email && touched.email}
      disabled={isSubmitting}
      helperText={
        errors.email ? errors.email : 'Enter your email address.'
      }
    />
    <TextField
      required
      id="password"
      label="Password"
      type="password"
      className={classes.textField}
      onChange={handleChange}
      onBlur={handleBlur}
      error={!!errors.password && touched.password}
      disabled={isSubmitting}
      helperText={
        errors.password
          ? errors.password
          : 'Enter a nice, long, unique password.'
      }
    />
    <Button
      variant="raised"
      color="primary"
      type="submit"
      className={classes.button}
    >
      {`${FORM_BUTTON_REGISTER} & ${FORM_BUTTON_LOGIN}`}
    </Button>
  </form>
);

RegisterForm.propTypes = {
  submitHandler: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
  classes: PropTypes.shape({
    textField: PropTypes.string.isRequired,
    button: PropTypes.string.isRequired
  }).isRequired,
  dispatch: PropTypes.func.isRequired,
  user: PropTypes.shape({
    error: PropTypes.string
  }),
  handleSubmit: PropTypes.func.isRequired,
  handleChange: PropTypes.func.isRequired,
  handleBlur: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool,
  values: PropTypes.shape({
    title: PropTypes.string,
    body: PropTypes.string,
    field_experience_path: PropTypes.string
  }).isRequired,
  errors: PropTypes.shape({
    title: PropTypes.oneOfType([PropTypes.string, PropTypes.bool]),
    field_experience_path: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.bool
    ]),
    body: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
  }).isRequired,
  touched: PropTypes.shape({
    title: PropTypes.bool,
    field_experience_path: PropTypes.bool,
    body: PropTypes.bool
  }).isRequired
};

RegisterForm.defaultProps = {
  user: {
    error: null
  }
};

const FormikRegisterForm = withFormik({
  displayName: 'RegisterForm',
  enableReinitialize: true,
  validationSchema: object().shape({
    username: string()
      .required()
      .min(3)
      .max(100)
      .matches(FORM_INPUT_VALIDATION_REGEX_URL_SEGMENT, {
        message:
          'Value must be URL-friendly. No spaces, no special characters, just letters, numbers, and dashes'
      }),
    email: string()
      .email()
      .required()
      .min(3)
      .max(100),
    password: string()
      .required()
      .min(3)
      .max(24)
  }),
  handleSubmit: (values, { props, setSubmitting }) => {
    const { user, dispatch } = props;
    const { username, email, password } = values;
    dispatch({
      type: USER_REGISTER,
      username,
      email,
      password
    });
  }
})(RegisterForm);

export default withStyles(RegisterFormStyles)(FormikRegisterForm);
