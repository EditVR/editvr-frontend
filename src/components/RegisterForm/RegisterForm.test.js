/**
 * @file RegisterForm.test.js
 * Contains tests for RegisterForm.js.
 */

import React from 'react';
import { mount, shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import RegisterForm from './RegisterForm.container';

describe('<RegisterForm />', () => {
  it('Matches its snapshot', () => {
    const store = configureStore()({
      user: {
        error: null
      }
    });

    expect(
      renderer.create(<RegisterForm store={store} />).toJSON()
    ).toMatchSnapshot();
  });

  it('Executes its submit handler as expected.', () => {
    const submitFn = jest.fn();
    const store = configureStore()({
      user: {
        error: null
      }
    });

    const wrapper = mount(
      shallow(
        shallow(<RegisterForm submitHandler={submitFn} store={store} />).get(0)
      ).get(0)
    );

    wrapper.find('form').simulate('submit');
    expect(submitFn).toHaveBeenCalledTimes(1);
  });
});
