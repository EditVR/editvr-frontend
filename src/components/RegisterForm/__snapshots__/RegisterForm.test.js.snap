// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RegisterForm /> Matches its snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <div
    aria-describedby="username-helper-text"
    className="MuiFormControl-root-3 C-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="username"
    >
      Username
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25"
        defaultValue={undefined}
        disabled={false}
        id="username"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="text"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="username-helper-text"
    >
      Enter a URL-friendly username.
    </p>
  </div>
  <div
    aria-describedby="email-helper-text"
    className="MuiFormControl-root-3 C-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="email"
    >
      Email
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25 MuiInput-inputType-28"
        defaultValue={undefined}
        disabled={false}
        id="email"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="email"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="email-helper-text"
    >
      Enter your email address.
    </p>
  </div>
  <div
    aria-describedby="password-helper-text"
    className="MuiFormControl-root-3 C-textField-1"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <label
      className="MuiFormLabel-root-12 MuiInputLabel-root-7 MuiInputLabel-formControl-8 MuiInputLabel-animated-11"
      data-shrink={false}
      htmlFor="password"
    >
      Password
      <span
        className="MuiFormLabel-asterisk-16"
      >
         *
      </span>
    </label>
    <div
      className="MuiInput-root-17 MuiInput-formControl-18 MuiInput-underline-21"
    >
      <input
        aria-invalid={false}
        aria-required={true}
        autoComplete={undefined}
        autoFocus={undefined}
        className="MuiInput-input-25 MuiInput-inputType-28"
        defaultValue={undefined}
        disabled={false}
        id="password"
        name={undefined}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        onKeyDown={undefined}
        onKeyUp={undefined}
        placeholder={undefined}
        readOnly={undefined}
        required={true}
        rows={undefined}
        type="password"
        value={undefined}
      />
    </div>
    <p
      className="MuiFormHelperText-root-30"
      id="password-helper-text"
    >
      Enter a nice, long, unique password.
    </p>
  </div>
  <button
    className="MuiButtonBase-root-50 MuiButton-root-34 MuiButton-raised-40 MuiButton-raisedPrimary-41 C-button-2"
    disabled={false}
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="submit"
  >
    <span
      className="MuiButton-label-35"
    >
      Register & Login
    </span>
    <span
      className="MuiTouchRipple-root-53"
    />
  </button>
</form>
`;
